### ISSUE TITLE
How to Train a YOLOv8 Model to Detect Custom Classes Along with COCO Dataset Classes

### ISSUE THREAD

**Issue by itstechaj** [ultralytics/ultralytics#3066]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

YOLOv8 is trained on the COCO dataset which includes 80 classes.
However, I want to create my own object detection model based on YOLOv8 that can detect additional classes not present in the COCO dataset. Currently, I am loading the pre-trained YOLOv8 object detection model and training it on my own custom dataset, which includes annotations for the additional classes I want to detect. However, after training, my model is only detecting the new annotated classes and not the original COCO dataset classes.
I'm new to deep learning, so please guide me on how to train my model to detect both the additional classes as well as the COCO dataset classes.

**Comment by bit-scientist** [comment]
You'll have to have the images of objects that you want to detect, namely, the entire COCO dataset. Then you put your dataset next to it and configure the `data.yaml` file as train, valid, test splits, with `nc` being 80 + `additional classes` .

Let's say you have a dataset called A (**datasetA**). All the images in **datasetA** must be properly annotated in a yolo-format. (I should note that the objects in 80 classes should also be labeled 100%.) Then you create a `data.yaml` file, something like below:

```
train:
 - D:\ultralytics\datasets\COCO\train\images
 - D:\ultralytics\datasets\datasetA\train\images

val:
 - D:\ultralytics\datasets\COCO\valid\images
 - D:\ultralytics\datasets\datasetA\valid\images

test:
 - D:\ultralytics\datasets\COCO\test\images
 - D:\ultralytics\datasets\datasetA\test\images

nc: 80 +  additional classes # the class names should be inserted below.

names: [ 'person', 'bicycle', 'car', 'motorcycle', ...] 

```

**Comment by itstechaj** [comment]
Thanks for answering.
So, if I understand correctly, I need to place my dataset alongside the COCO dataset and annotate all new images in my dataset with all the required classes (the existing 80 classes in COCO + new classes i want to include). Then, I will need to configure the YAML file. Finally, I can begin training the model with the whole dataset. Is that correct?

**Comment by itstechaj** [comment]
> You'll have to have the images of objects that you want to detect, namely, the entire COCO dataset. Then you put your dataset next to it and configure the `data.yaml` file as train, valid, test splits, with `nc` being 80 + `additional classes` .
> 
> Let's say you have a dataset called A (**datasetA**). All the images in **datasetA** must be properly annotated in a yolo-format. (I should note that the objects in 80 classes should also be labeled 100%.) Then you create a `data.yaml` file, something like below:
> 
> ```
> train:
>  - D:\ultralytics\datasets\COCO\train\images
>  - D:\ultralytics\datasets\datasetA\train\images
> 
> val:
>  - D:\ultralytics\datasets\COCO\valid\images
>  - D:\ultralytics\datasets\datasetA\valid\images
> 
> test:
>  - D:\ultralytics\datasets\COCO\test\images
>  - D:\ultralytics\datasets\datasetA\test\images
> 
> nc: 80 +  additional classes # the class names should be inserted below.
> 
> names: [ 'person', 'bicycle', 'car', 'motorcycle', ...] 
> ```

Thanks for answering.
So, if I understand correctly, I need to place my dataset alongside the COCO dataset and annotate all new images in my dataset with all the required classes (the existing 80 classes in COCO + new classes i want to include). Then, I will need to configure the YAML file. Finally, I can begin training the model with the whole dataset. Is that correct?

**Comment by bit-scientist** [comment]
@itstechaj yes, you don't actually have to put your dataset alongside COCO, whetever you put it, you show the path to it in yaml file. It's generally a good thing to store all datasets in `datasets` folder inside `ultralytics` project directory. Because some people (including myself) had some struggle to get it working when place in other places. So, in short, you're good to go.

**Comment by glenn-jocher** [comment]
Hi @bit-scientist,

To train your YOLOv8 object detection model to detect both the additional classes you want to include and the existing COCO dataset classes, you need to first annotate all the new images in your dataset with all the required classes (the existing 80 classes in COCO plus the new classes you want to include).

Then, you need to create a `data.yaml` file and specify the paths to the train, validation, and test sets along with the class names and the number of classes that your model should detect. Since you have additional classes, the number of classes that `nc` needs to be set to is 80 plus the number of additional classes.

It's generally recommended to store all datasets in the `datasets` folder inside the `ultralytics` project directory, and then you can reference the path to your dataset in the `data.yaml` file. After creating and configuring the `data.yaml` file, you can then start training your YOLOv8 object detection model with the whole dataset.

I hope this explanation helps you get started on training your model to detect both the COCO dataset classes and your additional classes. Let us know if you have any more questions or need further clarification.

**Comment by itstechaj** [comment]
[data.txt](https://github.com/ultralytics/ultralytics/files/11686718/data.txt)


> Hi @bit-scientist,
> 
> To train your YOLOv8 object detection model to detect both the additional classes you want to include and the existing COCO dataset classes, you need to first annotate all the new images in your dataset with all the required classes (the existing 80 classes in COCO plus the new classes you want to include).
> 
> Then, you need to create a `data.yaml` file and specify the paths to the train, validation, and test sets along with the class names and the number of classes that your model should detect. Since you have additional classes, the number of classes that `nc` needs to be set to is 80 plus the number of additional classes.
> 
> It's generally recommended to store all datasets in the `datasets` folder inside the `ultralytics` project directory, and then you can reference the path to your dataset in the `data.yaml` file. After creating and configuring the `data.yaml` file, you can then start training your YOLOv8 object detection model with the whole dataset.
> 
> I hope this explanation helps you get started on training your model to detect both the COCO dataset classes and your additional classes. Let us know if you have any more questions or need further clarification.

Should my custom dataset need to include images and labels for every class (80+ new class) that I want to detect? 

In my custom dataset's yaml file, I have added the digitalWhiteboard class at the 80th index position because, let's say, I want to detect an additional class (digitalWhiteboard) in addition to the 80 classes in the coco dataset. I have images of person and digitalWhiteboards only in my dataset, along with the corresponding class labels (0->person, 80->digitalWhiteboard). So after training, the Yolov8 model can easily detect person and digital whiteboards, but when I give it an image of a car, it is unable to detect the cars. Not only that, but if I give it an image of a dog or cat, then also it is not able to detect.

Please guide me on this.

I am attaching yaml file of my dataset as text here:-
[data.txt](https://github.com/ultralytics/ultralytics/files/11686821/data.txt)

**Comment by bit-scientist** [comment]
Your dataset, whether it's custom or COCO, must have objects of what you are eventually going to detect. In your example, there are persons and whiteboards in your dataset, after training you'll be able to detect only those. If you need cars (dogs, cats for that matter) to be detected as well, you MUST have images of those in your dataset as well.

**Comment by glenn-jocher** [comment]
@itstechaj hello! 

In order to detect specific objects with YOLOv8, your dataset needs to include images and labels for every class that you want the model to detect. In your example, your custom dataset includes images of persons and digital whiteboards, but in order to detect cars, cats, and dogs, you'll need to include images and labels for those objects as well. Without this data, the YOLOv8 model will not be able to detect those objects.

I hope this helps! Let us know if you have any further questions.

**Comment by itstechaj** [comment]
Thank you @bit-scientist @glenn-jocher now i got it.

**Comment by glenn-jocher** [comment]
@itstechaj you're welcome! We're happy to help. If you have any further questions or issues with YOLOv8, don't hesitate to ask.

**Comment by sanjaiiv04** [comment]
Hey, I am working on YOLOV8 semantic segmentation. I was checking with the pre-trained model 'yolov8s-seg.pt'. Now I want the model to only detect and segment certain classes in the COCO dataset and not all the classes. How to do that?
Note that I want to do it on Python not in CLI. I was able to filter the classes and begin detected through the CLI but I do not know how to do it in code.

**Comment by glenn-jocher** [comment]
@sanjaiiv04 hello,

You can customize the classes detected by the pre-trained model in your Python code by filtering the output of the inference. 

When you perform inference with the model, it will output detection predictions for each image, where each prediction includes the class, coordinates, and confidence score. 

You can then filter these predictions based on the class IDs of the classes you are interested in.

Here's a general pseudo-code breakdown:

1. Load the pre-trained model: Use the `torch.hub.load()` function.

2. Run inference: Use the `.predict()` function of the loaded model.

3. Filter the detections: After running inference, you'll get a list of detections for each image. Loop through these detections, and keep only those whose class ID matches the classes you are interested in.

This should help you get started with customizing the classes that YOLOv8 detects in your Python application. If you encounter any issues, don't hesitate to ask for further guidance.

**Comment by itstechaj** [comment]
> Hey, I am working on YOLOV8 semantic segmentation. I was checking with the pre-trained model 'yolov8s-seg.pt'. Now I want the model to only detect and segment certain classes in the COCO dataset and not all the classes. How to do that? Note that I want to do it on Python not in CLI. I was able to filter the classes and begin detected through the CLI but I do not know how to do it in code.

Hi @sanjaiiv04 
Python Code for your reference: -
I divided the whole code into three sections for better understanding.
```python
# <------------------------ Section 1 --------------------------------->
from ultralytics import YOLO

# Load a pretrained YOLOv8n model
model = YOLO('yolov8n.pt')

# Run inference on an image
results = model('traffic.jpg')  # results list

# desired_class ids -> this is a list of class ids that you want to consider (the id should be same as it is given in coco)
desired_class_ids=[0,2,3]
desired_class_names=dict(({0,"alpha"},{2,"gama"},{3,"delta"})) # class labels corresponding to desired_class_ids

# for checking the class_id corresponding to class_labels (for coco dataset) you can uncomment the below code
# print(results[0].names)

# <--------------------------- Section 2 ------------------------------------->
# Store desired results only
for r in results:
    # indexes -> list of indexes of only that classes which you want consider
    indexes = [index for index, item in enumerate(r.boxes.cls) if item in desired_class_ids]
    # now we are storing results for classes that you want to consider
    class_ids=r.boxes.cls[indexes] # stores the class_detected (includes only the classes that you want to consider)
    class_confs=r.boxes.conf[indexes]
    boxes_xyxy=r.boxes.xyxy[indexes] # stores the bounding boxes in xyxy format
    boxes_xyxyn=r.boxes.xyxyn[indexes] # stores the bounding boxes in xyxy (normalized) format
    boxes_xywh=r.boxes.xywh[indexes] # stores the bounding boxes in xywh format


# <----------------------------- Section 3 ---------------------------------->
# code for visualizing the results
import cv2
import os
from ultralytics.utils.plotting import Annotator,colors
def visualize_detections(img_src, output_dir="runs/", boxes_xyxy=None, class_ids=None, class_confs=None):
    """
    This function visualizes the detections by plotting bounding boxes on the image.

    Args:
    img_src: The input image path.
    output_dir (str, optional): The directory where the output will be saved. Defaults to "runs/".
    boxes_xyxy (list): The list of bounding boxes in the format (x1, y1, x2, y2).
    class_ids (list): The list of class IDs for each detected object.
    class_confs (list): The list of confidence scores for each detected object.

    Returns:
    numpy array: The image with plotted bounding boxes.
    """
    img=cv2.imread(img_src)
    # Create a copy of the image
    imgcpy = img.copy()
    
    # Initialize the annotator
    annotator = Annotator(imgcpy)
    
    # Loop over each detected object
    for i in range(len(class_ids)):
        # Add a bounding box and label to the image
        annotator.box_label(box=boxes_xyxy[i], label=(desired_class_names[class_ids[i].item()]+" "+ str(round(class_confs[0].item(), 2))), color=colors(class_ids[i], True))
    
    # Check if output directory exists
    if not os.path.exists(output_dir):
        # If not, create it
        os.makedirs(output_dir)
    
    # Save the image
    cv2.imwrite(os.path.join(output_dir, 'predicted_img.jpg'), annotator.result())
    # Return the annotated image
    return annotator.result()

# run the above function to visualize the predited boxes
visualize_detections("traffic.jpg",boxes_xyxy=boxes_xyxy,class_ids=class_ids,class_confs=class_confs)

# now your predicted image is saved at location runs/predicted_img.jpg
```
Output:-
![collage_final](https://github.com/ultralytics/ultralytics/assets/74065677/c3026c33-3395-4729-bae5-23ebe64ffef3)

Hope it helps you 👍.
**Feel free to discuss any further thoughts!**

**Comment by glenn-jocher** [comment]
Hi @itstechaj,

Sure, you can modify which classes your pre-trained model 'yolov8s-seg.pt' will detect and segment directly in your Python code. After you've loaded the model and performed inference on an image, you will receive an output of detection predictions for each image in your dataset. Each prediction includes several details, such as the class ID, bounding box coordinates, and a confidence score.

Here's where you can make your modifications: by filtering these predictions based on the class IDs. You simply need to create a condition that keeps only those predictions where the class ID matches the classes you want to detect (e.g., specific classes in the COCO dataset).

Remember, this operation takes place after performing inference, when going through the list of detections received as output.

Without providing code, please look into the 'results' object in the codebase where you'll find results for each object including 'class', 'box', and 'confidence'. Filter based on 'class' for your desired classes.

I hope this helps and happy coding!

**Comment by parth-patel2023** [comment]
I am using this code to train the yolov8n model where I change the path of the dataset and add one class after annotating images for only number-plate in yaml file 

`from paralytics import YOLO
 
# Load the model.
model = YOLO('yolov8n.pt')
 
# Training.
results = model.train(
   data='numberplate_v8.yaml',
   imgsz=640,
   epochs=10,
   batch=8,
   name='yolov8n_custom')`  
   
   but the final model does not detect cars or bikes (other classes than numberplate class)
   
   Am I making any mistake here?

**Comment by glenn-jocher** [comment]
@parth-patel2023 hello,

From your provided information, it seems like you've trained your model specifically on the 'numberplate' class, and thus, it has learned to detect only the 'numberplate'. It does not detect other classes like 'cars' or 'bikes' because those weren't included in your training dataset.

Suppose you want your model to detect other classes, like 'cars' or 'bikes'. In that case, you'll need to annotate images of these objects and include them in your training dataset. Ensure that these objects are present and properly labelled in your 'numberplate_v8.yaml' file.

In short, your model can only detect the classes it has been trained on. If it hasn't been trained on certain classes (like 'cars' or 'bikes'), it won't be able to identify those in images. To remedy this, ensure that your training dataset includes all the classes you want your model to detect and that these are all properly labelled.

I hope this clears up your issue. If anything is still unclear, please don't hesitate to ask further questions.

**Comment by parth-patel2023** [comment]
@glenn-jocher  
I get your point. I thought Yolov8n.pt already had other classes, like 'cars' or 'bikes', so when I add a numberplate class with annotated images, it will detect all classes, including the number plate.

This model used a transfer learning technique to train a new custom model for new classes. Does it go this way, or am I missing something?

**Comment by glenn-jocher** [comment]
Hello @parth-patel2023,

Your understanding of transfer learning is generally correct. When using transfer learning, we typically leverage the knowledge that a pre-trained model has learned(in this case, 'yolov8n.pt') and tune it for our specific application(in your case, for detecting number plates).

However, when fine-tuning a pre-trained model for a specific class or set of classes, you also need annotated images for those classes. If you only have annotated images for the number plate class, the model will specifically train to be good at detecting number plates. It won't be as effective for detecting classes that it has not seen during your fine-tuning phase, such as 'car' or 'bike', even though these classes might have been part of the original training data.

If you want the model to detect not only number plates but also other classes like 'car' or 'bike', you would also need to incorporate annotated images of these classes into your training data when you are fine-tuning the pre-trained model. This way, the model won't "forget" how to detect the other objects while learning to detect number plates better.

That's why even though you're applying transfer learning, the model didn't perform as you expected for the 'car' and 'bike' classes. If you add annotated images for these classes, you should see that your model begins detecting them as well.
  
Remember, the model is only as good as the data it's trained on - if it hasn't seen examples of certain classes during the training phase, it won't be able to effectively recognize these classes during inference.

Hope that clears things up! Let us know if there are any more questions.

**Comment by parth-patel2023** [comment]
@glenn-jocher 

Thank you for providing a detailed explanation of transfer learning and the importance of including annotated images for all relevant classes during fine-tuning. I appreciate your insights and your explanation has been very helpful in understanding the process better. Thank you!

I added the annotated images for 'car' and 'bikes' with numberplate class and trained a model with three classes, At the end, the custom model is accurate and fulfils my specific requirements. 

Amazing works in ultralytics!

**Comment by glenn-jocher** [comment]
@parth-patel2023,

You're most welcome! It's fantastic to hear that by including annotated images for 'cars' and 'bikes' alongside those for number plates, your custom YOLOv8 model is now accurately detecting all the desired classes. It's always rewarding to see your efforts pay off with a model that performs well and meets your specific needs.

Thank you for your kind words about our work at Ultralytics! We take great pride in developing tools that serve the community and help users like you achieve their goals in object detection. Your success is a testament to the robustness of the YOLOv8 model and the effectiveness of transfer learning for fine-tuning models on custom datasets.

If you have any more questions or need further assistance in the future, don't hesitate to reach out. Keep up the great work with your models, and we're excited for you to continue exploring the capabilities of YOLOv8.

Wishing you all the best with your projects!

The Ultralytics Team

**Comment by sanjaiiv04** [comment]
> > Hey, I am working on YOLOV8 semantic segmentation. I was checking with the pre-trained model 'yolov8s-seg.pt'. Now I want the model to only detect and segment certain classes in the COCO dataset and not all the classes. How to do that? Note that I want to do it on Python not in CLI. I was able to filter the classes and begin detected through the CLI but I do not know how to do it in code.
> 
> Hi @sanjaiiv04 Python Code for your reference (as suggested by @glenn-jocher): - I divided the whole code into three sections for better understanding.
> 
> ```python
> # <------------------------ Section 1 --------------------------------->
> from ultralytics import YOLO
> 
> # Load a pretrained YOLOv8n model
> model = YOLO('yolov8n.pt')
> 
> # Run inference on an image
> results = model('traffic.jpg')  # results list
> 
> # desired_class ids -> this is a list of class ids that you want to consider (the id should be same as it is given in coco)
> desired_class_ids=[0,2,3]
> desired_class_names=dict(({0,"alpha"},{2,"gama"},{3,"delta"})) # class labels corresponding to desired_class_ids
> 
> # for checking the class_id corresponding to class_labels (for coco dataset) you can uncomment the below code
> # print(results[0].names)
> 
> # <--------------------------- Section 2 ------------------------------------->
> # Store desired results only
> for r in results:
>     # indexes -> list of indexes of only that classes which you want consider
>     indexes = [index for index, item in enumerate(r.boxes.cls) if item in desired_class_ids]
>     # now we are storing results for classes that you want to consider
>     class_ids=r.boxes.cls[indexes] # stores the class_detected (includes only the classes that you want to consider)
>     class_confs=r.boxes.conf[indexes]
>     boxes_xyxy=r.boxes.xyxy[indexes] # stores the bounding boxes in xyxy format
>     boxes_xyxyn=r.boxes.xyxyn[indexes] # stores the bounding boxes in xyxy (normalized) format
>     boxes_xywh=r.boxes.xywh[indexes] # stores the bounding boxes in xywh format
> 
> 
> # <----------------------------- Section 3 ---------------------------------->
> # code for visualizing the results
> import cv2
> import os
> from ultralytics.utils.plotting import Annotator,colors
> def visualize_detections(img_src, output_dir="runs/", boxes_xyxy=None, class_ids=None, class_confs=None):
>     """
>     This function visualizes the detections by plotting bounding boxes on the image.
> 
>     Args:
>     img_src: The input image path.
>     output_dir (str, optional): The directory where the output will be saved. Defaults to "runs/".
>     boxes_xyxy (list): The list of bounding boxes in the format (x1, y1, x2, y2).
>     class_ids (list): The list of class IDs for each detected object.
>     class_confs (list): The list of confidence scores for each detected object.
> 
>     Returns:
>     numpy array: The image with plotted bounding boxes.
>     """
>     img=cv2.imread(img_src)
>     # Create a copy of the image
>     imgcpy = img.copy()
>     
>     # Initialize the annotator
>     annotator = Annotator(imgcpy)
>     
>     # Loop over each detected object
>     for i in range(len(class_ids)):
>         # Add a bounding box and label to the image
>         annotator.box_label(box=boxes_xyxy[i], label=(desired_class_names[class_ids[i].item()]+" "+ str(round(class_confs[0].item(), 2))), color=colors(class_ids[i], True))
>     
>     # Check if output directory exists
>     if not os.path.exists(output_dir):
>         # If not, create it
>         os.makedirs(output_dir)
>     
>     # Save the image
>     cv2.imwrite(os.path.join(output_dir, 'predicted_img.jpg'), annotator.result())
>     # Return the annotated image
>     return annotator.result()
> 
> # run the above function to visualize the predited boxes
> visualize_detections("traffic.jpg",boxes_xyxy=boxes_xyxy,class_ids=class_ids,class_confs=class_confs)
> 
> # now your predicted image is saved at location runs/predicted_img.jpg
> ```
> 
> Output:- ![collage_final](https://user-images.githubusercontent.com/74065677/279483348-c3026c33-3395-4729-bae5-23ebe64ffef3.jpg)
> 
> Hope it helps you 👍. **Feel free to discuss any further thoughts!**

@itstechaj This works! I was able to properly filter the classes detected thanks to you!!

**Comment by glenn-jocher** [comment]
@sanjaiiv04, it's great to hear that the guidance provided was helpful and you were able to filter the detections effectively in your Python code. It's always rewarding to achieve the desired results in your projects. If you encounter any further questions or require additional assistance as you continue to work with YOLOv8, please feel free to ask. Happy coding with your YOLOv8 semantic segmentation tasks!

**Comment by glenn-jocher** [comment]
@itstechaj if you ever need further clarification or assistance with your projects, don't hesitate to reach out to the Ultralytics/YOLOv8 community. Wishing you the best of success with your segmentation tasks!

**Comment by cobaltautomationdev** [comment]
> Hello @parth-patel2023,
> 
> Your understanding of transfer learning is generally correct. When using transfer learning, we typically leverage the knowledge that a pre-trained model has learned(in this case, 'yolov8n.pt') and tune it for our specific application(in your case, for detecting number plates).
> 
> However, when fine-tuning a pre-trained model for a specific class or set of classes, you also need annotated images for those classes. If you only have annotated images for the number plate class, the model will specifically train to be good at detecting number plates. It won't be as effective for detecting classes that it has not seen during your fine-tuning phase, such as 'car' or 'bike', even though these classes might have been part of the original training data.
> 
> If you want the model to detect not only number plates but also other classes like 'car' or 'bike', you would also need to incorporate annotated images of these classes into your training data when you are fine-tuning the pre-trained model. This way, the model won't "forget" how to detect the other objects while learning to detect number plates better.
> 
> That's why even though you're applying transfer learning, the model didn't perform as you expected for the 'car' and 'bike' classes. If you add annotated images for these classes, you should see that your model begins detecting them as well.
> 
> Remember, the model is only as good as the data it's trained on - if it hasn't seen examples of certain classes during the training phase, it won't be able to effectively recognize these classes during inference.
> 
> Hope that clears things up! Let us know if there are any more questions.

Is it possible that we could simply add a parameter "keep_original_class = True" in our future models and train the YOLO algorithm on our new classes? It seems like a waste of time to have to annotate images again for trained classes. Although, I must admit, this idea of mine may just be wishful thinking.

**Comment by parth-patel2023** [comment]
> > Hello @parth-patel2023,
> > Your understanding of transfer learning is generally correct. When using transfer learning, we typically leverage the knowledge that a pre-trained model has learned(in this case, 'yolov8n.pt') and tune it for our specific application(in your case, for detecting number plates).
> > However, when fine-tuning a pre-trained model for a specific class or set of classes, you also need annotated images for those classes. If you only have annotated images for the number plate class, the model will specifically train to be good at detecting number plates. It won't be as effective for detecting classes that it has not seen during your fine-tuning phase, such as 'car' or 'bike', even though these classes might have been part of the original training data.
> > If you want the model to detect not only number plates but also other classes like 'car' or 'bike', you would also need to incorporate annotated images of these classes into your training data when you are fine-tuning the pre-trained model. This way, the model won't "forget" how to detect the other objects while learning to detect number plates better.
> > That's why even though you're applying transfer learning, the model didn't perform as you expected for the 'car' and 'bike' classes. If you add annotated images for these classes, you should see that your model begins detecting them as well.
> > Remember, the model is only as good as the data it's trained on - if it hasn't seen examples of certain classes during the training phase, it won't be able to effectively recognize these classes during inference.
> > Hope that clears things up! Let us know if there are any more questions.
> 
> @glenn-jocher Is it possible that we could simply add a parameter "keep_original_class = True" in our future models and train the YOLO algorithm on our new classes? It seems like a waste of time to have to annotate images again for trained classes. Although, I must admit, this idea of mine may just be wishful thinking.

@cobaltautomationdev  @glenn-jocher 
if this is applicable then it's very helpful and time effective.

**Comment by glenn-jocher** [comment]
Hello @parth-patel2023 and @parth-patel2023,

The concept of a "keep_original_classes=True" parameter is an interesting one and touches upon the core principles of transfer learning and incremental learning. However, the way these learning paradigms are currently implemented generally requires that the model sees examples of all the classes it needs to detect during training. This is to prevent catastrophic forgetting, where a neural network forgets how to perform tasks it was previously trained on when it learns new tasks.

In practice, mixing old and new classes without providing any old class examples would require the model to retain its previous knowledge (in this case, the COCO classes) while learning to detect new classes (like number plates). It's a challenging problem, but one potential avenue to achieve this might involve techniques from continual learning or using a "regularization term" during training that helps to retain old knowledge. 

Such a feature as "keep_original_classes=True" is not trivial to implement because the model's ability to generalize from its learned representations isn't perfect, especially when the new data significantly differs from the old. Training with both old and new classes typically balances the model's performance across all classes. 

In practical terms, the current best practice is indeed to retrain the model with a mix of both the new custom dataset and samples from the original dataset. For now, although it may be not ideal in terms of efficiency, it's the proven method to ensure the model performs well on both the original and new classes.

We always welcome ideas from the community, as they often lead to new features and innovations. Your feedback is valuable and may inspire future development directions that make the training process even more efficient.

Keep up the great work, and please continue sharing your thoughts and suggestions!

**Comment by glenn-jocher** [comment]
@Dimasilham7 hi there,

It sounds like there might be an issue with how the classes are defined or read from your YAML file. Please ensure that your YAML file is correctly formatted and that the class names are properly listed under the `names` field. Also, verify that the class indices in your annotations match the class names in the YAML file. If the issue persists, please check for any discrepancies between your dataset and the YAML configuration. If you need further assistance, feel free to share more details or the YAML file content for a closer look.

**Comment by Dimasilham7** [comment]
> @Dimasilham7 hi there,
> 
> It sounds like there might be an issue with how the classes are defined or read from your YAML file. Please ensure that your YAML file is correctly formatted and that the class names are properly listed under the `names` field. Also, verify that the class indices in your annotations match the class names in the YAML file. If the issue persists, please check for any discrepancies between your dataset and the YAML configuration. If you need further assistance, feel free to share more details or the YAML file content for a closer look.

hi, actually, this my yaml file result


`path: ../datasets/roboflow

train: train/images
val: valid/images
test: test/images

names: 
  0: Body Harness
  1: Cat
  2: Chicken
  3: Cow
  4: Fire
  5: Gloves
  6: Goggle
  7: Hardhat
  8: NO-Hardhat
  9: NO-Safety Shoes
  10: NO-Safety Vest
  11: No-Gloves
  12: No-Goggle
  13: Person
  14: Puddle
  15: SUV
  16: Shoes
  17: Smoke`
  
 I'm annotating using Roboflow, but in this case, the class name is like this, the question is do I still need this YAML file? Or do I need to make the name into the list of classes like this when I'm trained?
  
  `class = ['Smoke', 'Shoes']`

**Comment by glenn-jocher** [comment]
@Dimasilham7, your YAML file should list class names as an array, not as a key-value mapping. Here's the correct format:

```yaml
names:
  - Body Harness
  - Cat
  - Chicken
  # ... other classes
  - Smoke
```

Ensure your YAML reflects this structure. If you're using Roboflow, it typically handles class names for you, but you'll still need the YAML file to define the dataset paths and class names for training with YOLOv8. No need to manually define a `class` list in your training script if the YAML is correct.

**Comment by Dimasilham7** [comment]
> @Dimasilham7, your YAML file should list class names as an array, not as a key-value mapping. Here's the correct format:
> 
> ```yaml
> names:
>   - Body Harness
>   - Cat
>   - Chicken
>   # ... other classes
>   - Smoke
> ```
> 
> Ensure your YAML reflects this structure. If you're using Roboflow, it typically handles class names for you, but you'll still need the YAML file to define the dataset paths and class names for training with YOLOv8. No need to manually define a `class` list in your training script if the YAML is correct.

@glenn-jocher since the YAML file it's produce from roboflow, if the YAML file result is wrong, do i need to alter it? and make a new one that fit with YAML file yolov8

**Comment by glenn-jocher** [comment]
@Dimasilham7, if the YAML file generated by Roboflow doesn't match the expected format for YOLOv8, you should modify it to follow the correct structure. Ensure the `names` field is a list of class names as shown in the example. Once corrected, use this updated YAML file for training your YOLOv8 model.

**Comment by Y-T-G** [comment]
I wrote a guide on how to do this using an extra head trick:
https://y-t-g.github.io/tutorials/yolov8n-add-classes/

**Comment by glenn-jocher** [comment]
@Y-T-G hey there! 😊

Thanks for sharing your guide on adding extra classes to YOLOv8 using the extra head trick. It looks like a valuable resource for the community! If anyone's looking to expand their YOLOv8 model with additional classes, definitely check it out. 🚀

```yaml
# Example snippet from the guide
head:
  - from: [head]
    number: [num_classes + new_classes]  # Adjust class count
    ...
```

Keep up the great work!

**Comment by elsa-sachin** [comment]
@glenn-jocher Hi ,So am working on a project to detect the defects found in a construction site using the pre-trained yolov8n.pt model. Right now am dealing with one dataset that is GLASS with 2 classes {0:Glass_damage,1:Glass_not_damage}.
Now i would like to add a new dataset of switches which has the same class indices ,i tried transfer learning ,used the best.pt of glass to train on the switch dataset . I tired 2 methods freezing and adding a new head.Nothing is working.As both the datasets have same class indices ,i think it may overwrite i dont want that to happen .Can help me with this?

**Comment by glenn-jocher** [comment]
Hi @elsa-sachin,

Thank you for reaching out! It sounds like you're working on an interesting project. Combining datasets with overlapping class indices can indeed be challenging. Here are a few steps you can take to address this issue:

1. **Reindex Classes**: Ensure that the class indices for your new dataset (switches) do not overlap with your existing dataset (glass). You can reindex the classes in your new dataset to avoid conflicts. For example, if your glass dataset has classes {0: Glass_damage, 1: Glass_not_damage}, you could reindex your switch dataset to use classes {2: Switch_damage, 3: Switch_not_damage}.

2. **Merge Datasets**: Combine your datasets into a single dataset with unique class indices. Update your YAML configuration file to reflect the new class structure. Here’s an example of how your updated YAML file might look:

    ```yaml
    path: ../datasets/combined
    train: images/train
    val: images/val
    test: images/test

    names:
      0: Glass_damage
      1: Glass_not_damage
      2: Switch_damage
      3: Switch_not_damage
    ```

3. **Retrain the Model**: Train your model on the combined dataset with the updated class indices. This ensures that the model learns to distinguish between all classes across both datasets.

    ```python
    from ultralytics import YOLO

    # Load a pretrained model
    model = YOLO("yolov8n.pt")

    # Train the model on the combined dataset
    results = model.train(data="path/to/combined_dataset.yaml", epochs=100, imgsz=640)
    ```

4. **Verify and Test**: After training, verify the model's performance on both datasets to ensure it correctly identifies all classes.

If you encounter any issues during this process, please provide a reproducible example so we can better assist you. You can find guidance on creating a minimum reproducible example [here](https://docs.ultralytics.com/help/minimum_reproducible_example).

Additionally, make sure you are using the latest versions of the Ultralytics packages to benefit from the latest features and bug fixes.

Feel free to reach out if you have any further questions. Good luck with your project! 🚀

**Comment by elsa-sachin** [comment]
Thank you so much for the reply!  So I am using **Roboflow** to annotate my dataset. So for an instance, my first dataset was GLASS with 2 classes {0,1} I annotated and downloaded the dataset along with the labels folder and the data.yaml file.Now, if I had to add other datasets such as switches,wall with 2 classes each respectively ,I would have to add it again into my glass dataset in robflow annotate the glass ,wall and switches dataset again ,download the new yaml file and train it. So wouldn't this process continue if I keep adding new datasets and have to train the model every time and wouldn't it take so much time. You did mention reindexing the class labels,even if I reindexed the labels manually in my yaml file ,the labels of the images are not gonna change and it would get confused.For an example,Glass images has labels {0,1} and similarly for switch{0,1} .But in the yaml file it would {0,1,2,3} where {0,1} is for glass and {2,3}is for switches. I hope you did understand my challenges and help me with it .
Thanking you in advance
